core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -341,20 +344,54 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
         final RoutingResult routingResult = routed.routingResult();
         final ServiceConfig serviceCfg = routed.value();
         final HttpService service = serviceCfg.service();
-
+        final EventLoop serviceEventLoop;
+        final EventLoopGroup serviceWorkerGroup = serviceCfg.serviceWorkerGroup();
+        if (serviceWorkerGroup == config.workerGroup()) {
+            serviceEventLoop = eventLoop;
+        } else {
+            serviceEventLoop = serviceWorkerGroup.next();
+        }
         final DefaultServiceRequestContext reqCtx = new DefaultServiceRequestContext(
-                serviceCfg, channel, config.meterRegistry(), protocol,
+                serviceCfg, channel, serviceEventLoop, config.meterRegistry(), protocol,
                 nextRequestId(), routingCtx, routingResult, req.exchangeType(),
                 req, sslSession, proxiedAddresses, clientAddress,
                 req.requestStartTimeNanos(), req.requestStartTimeMicros());
 
+        final boolean needsDirectExecution = serviceEventLoop == config.workerGroup();
+
+        if (needsDirectExecution) {
+            serveRequest(ctx, req, serviceCfg, service, routingCtx, routingResult,
```

Comment:
How about reverting this change?
The only thing the `serviceEventLoop` is invoking `service.serve(ctx, reqCtx);`.
But this code delegate everything to the `serviceEventLoop`. 
Here's my original thought about using different `eventLoop`:
```java
try (SafeCloseable ignored = reqCtx.push()) {
    final HttpResponse serviceResponse;
    if (needsDirectExecution) {
        serviceResponse = serve(req, service, reqCtx);
    } else {
        serviceResponse = HttpResponse.from(
                () -> {
                    try (SafeCloseable ignored2 = reqCtx.push()) {
                        return serve(req, service, reqCtx);
                    }
                }, serviceEventLoop);
    }

    // Keep track of the number of unfinished requests and
    // clean up the request stream when response stream ends.
    final boolean isTransientService =
            serviceCfg.service().as(TransientService.class) != null;
    if (!isTransientService) {
        gracefulShutdownSupport.inc();
    }
    unfinishedRequests.put(req, serviceResponse);
    ... // below code are the same as the current master
}

private HttpResponse serve(DecodedHttpRequest req, HttpService service,
                           DefaultServiceRequestContext reqCtx) {
    final ServerErrorHandler serverErrorHandler = config.errorHandler();
    HttpResponse serviceResponse;
    try {
        req.init(reqCtx);
        serviceResponse = service.serve(reqCtx, req);
    } catch (Throwable cause) {
        // No need to consume further since the response is ready.
        if (cause instanceof HttpResponseException || cause instanceof HttpStatusException) {
            req.close();
        } else {
            req.close(cause);
        }
        serviceResponse = HttpResponse.ofFailure(cause);
    }

    serviceResponse = serviceResponse.recover(cause -> {
        // Store the cause to set as the log.responseCause().
        CapturedServiceException.set(reqCtx, cause);
        // Recover the failed response with the error handler.
        return serverErrorHandler.onServiceException(reqCtx, cause);
    });
    return serviceResponse;
}


```